### PR TITLE
[Gekidou MM-3972] Websocket Events - Reaction and emojis

### PR DESCRIPTION
**Pull request by jfrerich**
#### Summary
This PR handles the following teams websocket events:
* Reaction added event 
* Reaction removed event
* Emoji added event

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-41523

#### Checklist

- [ ] Added or updated unit tests (required for all new features)
- [ ] Has UI changes
- [ ] Includes text changes and localization file updates
- [ ] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
This PR was tested on: 

#### Release Note

```release-note

```

### CODE REVIEW

File: app/database/operator/server_data_operator/handlers/user.ts [hunk 1]
```diff
@@ -165,6 +166,7 @@ const UserHandler = (superclass: any) => class extends superclass {
             }
 
             if (deleteReactions?.length && !skipSync) {
+                deleteReactions.map((outCast) => outCast.prepareDestroyPermanently());
```
Comment (enahum) [inline app/database/operator/server_data_operator/handlers/user.ts:169]: this should not be a map, this should be a `forEach`

Comment (enahum) [reply app/database/operator/server_data_operator/handlers/user.ts:169]:
```suggestion
                deleteReactions.forEach((outCast) => outCast.prepareDestroyPermanently());
```

File: app/actions/websocket/index.ts [hunk 2]
```diff
@@ -260,17 +261,17 @@ export async function handleEvent(serverUrl: string, msg: WebSocketMessage) {
...
             break;
 
-        // return dispatch(handleReactionAddedEvent(msg));
         case WebsocketEvents.REACTION_REMOVED:
+            handleReactionRemovedFromPostEvent(serverUrl, msg);
             break;
 
-        // return dispatch(handleReactionRemovedEvent(msg));
         case WebsocketEvents.EMOJI_ADDED:
+            handleAddCustomEmoji(serverUrl, msg);
```
Comment (enahum) [inline app/actions/websocket/index.ts:272]: do we have an `EMOJI_REMOVED` event?

File: app/actions/websocket/reactions.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+export async function handleAddCustomEmoji(serverUrl: string, msg: WebSocketMessage): Promise<void> {
+    const operator = DatabaseManager.serverDatabases[serverUrl]?.operator;
+    if (!operator) {
+        return;
+    }
+
+    try {
+        const emoji = JSON.parse(msg.data.emoji) as CustomEmoji;
+        if (operator) {
```
Comment (enahum) [inline app/actions/websocket/reactions.ts:17]: no need to check for this as it would return on line 12

File: app/actions/websocket/reactions.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    } catch {
+        // Do nothing
+    }
+}
+
+export async function handleReactionAddedToPostEvent(serverUrl: string, msg: WebSocketMessage): Promise<void> {
+    const database = DatabaseManager.serverDatabases[serverUrl];
+    if (!database) {
+        return;
+    }
```
Comment (enahum) [inline app/actions/websocket/reactions.ts:32]: const operator = DatabaseManager.serverDatabases[serverUrl]?.operator;
    if (!operator) {
        return;
    }

File: app/actions/websocket/reactions.ts [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+    try {
+        const reaction = JSON.parse(msg.data.reaction) as Reaction;
+        const operator = database?.operator;
+        if (operator) {
```
Comment (enahum) [inline app/actions/websocket/reactions.ts:37]: no need for this after the suggestion I made

File: app/actions/websocket/reactions.ts [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+export async function handleAddCustomEmoji(serverUrl: string, msg: WebSocketMessage): Promise<void> {
+    const operator = DatabaseManager.serverDatabases[serverUrl]?.operator;
+    if (!operator) {
+        return;
+    }
+
+    try {
+        const emoji = JSON.parse(msg.data.emoji) as CustomEmoji;
+        if (operator) {
+            operator.handleCustomEmojis({
```
Comment (enahum) [inline app/actions/websocket/reactions.ts:18]: if you don't await the error won't be caught

**Review by enahum — CHANGES_REQUESTED** [review]
interesting to see, I'm almost sure we have a WS event for when a custom emoji is removed, perhaps it was never added to the mobile app, but we could do now!

**Comment by jfrerich** [comment]
We do not have Websocket event from the server when a emoji is deleted. See [emoji.go](https://github.com/mattermost/mattermost-server/blob/master/app/emoji.go#L169)

I could add code the WS handler in this PR and and create the server PR also or add as a new ticket to get this merged sooner.

**Comment by enahum** [comment]
Let's create a ticket for the missing custom emoji removal

**Comment by enahum** [comment]
@jfrerich did you forget to push a commit ?

**Comment by jfrerich** [comment]
@enahum, yes, I forgot to push. Sorry. It's there now.

File: app/actions/websocket/reactions.ts [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+            prepareRecordsOnly: false,
+            emojis: [emoji],
+        });
+    } catch {
+        // Do nothing
+    }
+}
+
+export async function handleReactionAddedToPostEvent(serverUrl: string, msg: WebSocketMessage): Promise<void> {
+    console.log('\n<><> 2. handleReactionAddedToPostEvent');
```
Comment (enahum) [inline app/actions/websocket/reactions.ts:27]: remove!
```suggestion
```

File: app/actions/websocket/reactions.ts [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+                reactions: [reaction],
+            }],
+        });
+    } catch {
+        // Do nothing
+    }
+}
+
+export async function handleReactionRemovedFromPostEvent(serverUrl: string, msg: WebSocketMessage): Promise<void> {
+    const database = DatabaseManager.serverDatabases[serverUrl].database;
```
Comment (enahum) [inline app/actions/websocket/reactions.ts:49]:
```suggestion
    const database = DatabaseManager.serverDatabases[serverUrl]?.database;
```

**Review by enahum — CHANGES_REQUESTED** [review]

**Review by enahum — APPROVED** [review]

**Review by larkox — APPROVED** [review]
LGTM!

**Pull request merged**